engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java

```diff
@@ -159,9 +160,17 @@ private void verifyIncidentResolving(final BpmnElementContext context, final Str
       endOfExecutionPath = element.getOutgoing().isEmpty();
     }
 
+    final AtomicBoolean satisfiesCompletionCondition = new AtomicBoolean();
     if (endOfExecutionPath) {
       final var failureOrSuccess =
-          beforeExecutionPathCompleted(element, context).map(ok -> context);
+          beforeExecutionPathCompleted(element, context)
+              .map(
+                  ok -> {
+                    if (ok instanceof Boolean) {
+                      satisfiesCompletionCondition.set((Boolean) ok);
+                    }
+                    return context;
+                  });
       if (failureOrSuccess.isLeft()) {
         return failureOrSuccess;
       }
```

Comment:
🔧 No need for an AtomicBoolean here. I guess this boils down to personal opinion, but I'd prefer this something more like this:
```java
final Either<Failure, ?> satisfiesCompletionConditionOrFailure;
if (endOfExecutionPath) {
  satisfiesCompletionConditionOrFailure = beforeExecutionPathCompleted(element, context);
  if (satisfiesCompletionConditionOrFailure.isLeft()) {
    return satisfiesCompletionConditionOrFailure.map(ok -> context);
  }
} else {
  satisfiesCompletionConditionOrFailure = Either.right(null);
}

final var completed = transitionTo(context, ProcessInstanceIntent.ELEMENT_COMPLETED);
metrics.elementInstanceCompleted(context);

if (endOfExecutionPath) {
  // afterExecutionPathCompleted is not allowed to fail (incident would be unresolvable)
  afterExecutionPathCompleted(element, completed, satisfiesCompletionConditionOrFailure.get());
}
return Either.right(completed);
```

TBH, the more I look at this code, the more I think we can further clean this up by using the Either type as the control flow instead of the if statements `x.isLeft()` and such. But that's out of scope for now.